Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1381 @@
+  * What is the monad that should be used and its canonical operations with `%monad` (we reuse `Alex` for simplicity).
+  * Which function lexes tokens and how to extract the EOF token with `%lexer`.
+* `%%`, after which we must specify the production rules for the grammar.
+* A trailer to include more Haskell definitions, such as auxiliary files and the AST that will be built by the parser. You can also define those in another module if you prefer.
+
+### Tokens
+
+Since Happy operates on tokens, we must instruct it on how to match each token.
+This is done using `%token`, which must be placed before the `%%` in the file.
+For each token, on the left side, we'll put an alias that we can use to refer to our terminals.
+On the right side, we use a Haskell pattern between braces indicating how Happy should match the tokens we've defined while working with Alex.
+
+```alex
+%token
+  -- Identifiers
+  identifier { L.RangedToken (L.Identifier _) _ }
+  -- Constants
+  string     { L.RangedToken (L.String _) _ }
+  integer    { L.RangedToken (L.Integer _) _ }
+  -- Keywords
+  let        { L.RangedToken L.Let _ }
+  in         { L.RangedToken L.In _ }
+  if         { L.RangedToken L.If _ }
+  then       { L.RangedToken L.Then _ }
+  else       { L.RangedToken L.Else _ }
+  -- Arithmetic operators
+  '+'        { L.RangedToken L.Plus _ }
+  '-'        { L.RangedToken L.Minus _ }
+  '*'        { L.RangedToken L.Times _ }
+  '/'        { L.RangedToken L.Divide _ }
+  -- Comparison operators
+  '='        { L.RangedToken L.Eq _ }
+  '<>'       { L.RangedToken L.Neq _ }
+  '<'        { L.RangedToken L.Lt _ }
+  '<='       { L.RangedToken L.Le _ }
+  '>'        { L.RangedToken L.Gt _ }
+  '>='       { L.RangedToken L.Ge _ }
+  -- Logical operators
+  '&'        { L.RangedToken L.And _ }
+  '|'        { L.RangedToken L.Or _ }
+  -- Parenthesis
+  '('        { L.RangedToken L.LPar _ }
+  ')'        { L.RangedToken L.RPar _ }
+  -- Lists
+  '['        { L.RangedToken L.LBrack _ }
+  ']'        { L.RangedToken L.RBrack _ }
+  ','        { L.RangedToken L.Comma _ }
+  -- Types
+  ':'        { L.RangedToken L.Colon _ }
+  '->'       { L.RangedToken L.Arrow _ }
+```
+
+These tokens are the terminals used while writing the grammar. They are first-class, and we can use them among other symbols.
+
+### Semantic actions
+
+Now that we have the definitions of our tokens, we can now begin writing the parser itself.
+We can start with top-level definitions of format `let example = 0`.
+This will be a simplistic definition for now, but we will elaborate on it shortly.
+Note that production rules must be placed after the `%%` in the file, and you may delete the `empty` production after inserting the one below.
+
+```happy
+dec
+  : let identifier '=' integer {}
+```
+
+We use `:` to describe the production to be parsed (called `dec`), denoting the tokens (defined above) to be parsed, separated by spaces.
```

Comment:
![image](https://user-images.githubusercontent.com/30162364/173370509-f7ff7835-3077-4853-a770-c248ec53ad1f.png)

I think it depends on how pedantic we want to get with the terminology. A production is a substitution of form `a -> b`. `a` is the production header to which we can refer to the non-terminal described here. `b` is the production body, which groups symbols together. To be honest, I'm still a bit confused by this, by the terminology, but seems like it's non-terminal, changed it.

To be honest, I'm also a bit confused with the terminology. @lierdakil perhaps can help clear the confusion?

Reply:
Production head isn't necessarily a single non-terminal. This is true for context-free grammars, which we consider here, but not for context-dependent ones. Hence the distinction. It also bears mentioning that a single production head can (and usually will) have multiple corresponding production rules (with different bodies), so referring to "a production" by its head is ambiguous.

In this here example, "Exp", "Exp1", "Term" and "Factor" are non-terminals, "let", "var", etc are terminals, and "Exp : let var '=' Exp in Exp" is a single production rule, as is "Exp : Exp1", or "Term : Factor".

Reply:
So my understanding is the following:
* In Happy, a production head and a non-terminal refer to the same thing, as it has form `a : b` instead of `a b c : d` (which is a CSG), for example.
* A production with form `a : b | c` is the same as having two production rules `a : b` and `a : c`. And so, each pair with the head and body forms a production rule. Furthermore, `b` and `c` are also individually referred to as the bodies of `a` (note the plural).
* In Happy, it's not correct to refer to the production body as a non-terminal, and therefore, I should have said here that "We use `:` to describe the production body to be parsed (whose production head is `dec`)." These two in conjunction form a production [rule]. (I changed to this phrasing BTW)

Reply:
> In Happy, a production head and a non-terminal refer to the same thing

Arguably, yes. The insufferable pedant inside my head says these are distinct entities, but practically in a CFG there's no difference between a non-terminal and a production head, every production head is a non-terminal and vice versa (otherwise the grammar would be incomplete)

> A production with form a : b | c is the same as having two production rules a : b and a : c

Yes, at least from the theoretical standpoint. IIRC, Happy doesn't handle the latter syntax, however.

>  each pair with the head and body forms a production rule

 Yes.

>  Furthermore, b and c are also individually referred to as the bodies of a (note the plural).

Technically, bodies of productions with head `a`, but this is somewhat useless pedantry.

> it's not correct to refer to the production body as a non-terminal

This is true regardless of Happy (unless a given production body consists of a single non-terminal, which is somewhat rare).